src/Temporalio/Worker/WorkflowInstance.cs

```diff
@@ -422,51 +488,229 @@ private void Apply(WorkflowActivationJob job)
+            {
+                if (!bufferedSignals.TryGetValue(signal.SignalName, out var signalList))
+                {
+                    signalList = new();
+                    bufferedSignals[signal.SignalName] = signalList;
+                }
+                signalList.Add(signal);
+                return;
+            }
+
+            // Run the handler as a top-level function
+            _ = QueueNewTaskAsync(() => RunTopLevelAsync(async () =>
+            {
+                await inbound.Value.HandleSignalAsync(new(
+                    Signal: signal.SignalName,
+                    Definition: signalDefn,
+                    Args: DecodeArgs(
+                        signalDefn.Method ?? signalDefn.Delegate!.Method,
+                        signal.Input,
+                        $"Signal {signal.SignalName}"),
+                    Headers: signal.Headers)).ConfigureAwait(true);
+            }));
+        }
+
+        private void ApplyStartWorkflow(StartWorkflow start)
+        {
+            _ = QueueNewTaskAsync(() => RunTopLevelAsync(async () =>
+            {
+                var input = new ExecuteWorkflowInput(
                     Instance: Instance,
-                    RunMethod: details.Definition.RunMethod,
-                    Parameters: StartParameters,
-                    Headers: details.Start.Headers)).ConfigureAwait(false);
-                var result = details.PayloadConverter.ToPayload(resultObj);
+                    RunMethod: defn.RunMethod,
+                    Args: startArgs!.Value,
+                    Headers: start.Headers);
+                // We no longer need start args after this point, so we are unsetting them
+                startArgs = null;
+                var resultObj = await inbound.Value.ExecuteWorkflowAsync(input).ConfigureAwait(true);
+                var result = payloadConverter.ToPayload(resultObj);
                 AddCommand(new() { CompleteWorkflowExecution = new() { Result = result } });
             }));
         }
 
+        private void ApplyUpdateRandomSeed(UpdateRandomSeed update) =>
+            Random = new(unchecked((int)update.RandomnessSeed));
+
+        private void OnQueryDefinitionAdded(string name, WorkflowQueryDefinition defn)
+        {
+            if (name != defn.Name)
+            {
+                throw new ArgumentException($"Query name {name} doesn't match definition name {defn.Name}");
+            }
+        }
+
+        private void OnSignalDefinitionAdded(string name, WorkflowSignalDefinition defn)
+        {
+            if (name != defn.Name)
+            {
+                throw new ArgumentException($"Signal name {name} doesn't match definition name {defn.Name}");
+            }
+            // Send all buffered signals
+            if (bufferedSignals.TryGetValue(name, out var buffered))
+            {
+                bufferedSignals.Remove(name);
+                buffered.ForEach(ApplySignalWorkflow);
```

Comment:
Can you confirm that these signals get a chance to be processed before the workflow execute function is resumed so no signals are lost in case the workflow attempts to complete?

Reply:
Not following the question exactly. `OnSignalDefinitionAdded` is for when a user adds a signal handler manually and we are draining these inline at that time. You are already running the workflow code when you're adding a signal handler manually, there is no separate resumption.

Reply:
So any signal handler that I add dynamically is immediately invoked?
In this example will the workflow be prevented from completing?

(pseudocode)
```
bool preventFromCompleting = false;
Workflow.Signals.Add(BlaSignal, () =>
{
  preventFromCompleting = true;
});

if (preventFromCompleting)
{
  // do something
}
```

Reply:
Correct